File: InterviewPreparationKit/RecursiveStaircase.py
#!/bin/python3
# https://www.hackerrank.com/challenges/ctci-recursive-staircase/problem
import math
import os
import random
import re
import sys

# Complete the stepPerms function below.
def stepPerms(n):
    if n < 3:
        return n

    a = 1
    b = 2
    c = 4

    for i in range(3, n):
        current = a + b + c
        a = b
        b = c
        c = current

    return c


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    s = int(input())

    for s_itr in range(s):
        n = int(input())

        res = stepPerms(n)

        fptr.write(str(res) + '\n')

    fptr.close()

File: InterviewPreparationKit/RecursiveStaircase.java
// https://www.hackerrank.com/challenges/ctci-recursive-staircase/problem
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    // Complete the stepPerms function below.
    static int stepPerms(int n) {
        if (n < 3) {
            return n;
        }

        int[] array = new int[3];
        array[0] = 1;
        array[1] = 2;
        array[2] = 4;
        int current;

        for (int i = 3; i < n; i++) {
            current = array[2] + array[1] + array[0];
            array[0] = array[1];
            array[1] = array[2];
            array[2] = current;
        }
        
        return array[2];
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int s = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int sItr = 0; sItr < s; sItr++) {
            int n = scanner.nextInt();
            scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

            int res = stepPerms(n);

            bufferedWriter.write(String.valueOf(res));
            bufferedWriter.newLine();
        }

        bufferedWriter.close();

        scanner.close();
    }
}
